networkit/cpp/centrality/ComplexPaths.cpp

```diff
@@ -0,0 +1,203 @@
+    return complexPathGraph;
+}
+
+vector<node> ComplexPathAlgorithm::getAdopters() {
+    if (mode != Mode::singleNode)
+        WARN("complexPathAlgorithm[getAdopters]: no results in Mode::allNodes.");
+    assureFinished();
+    return adopters;
+}
+
+static void addNewEdge(Graph &g, node u, node v) {
+    if (g.hasEdge(u, v))
+        return;
+    g.addEdge(u, v);
+}
+
+Graph ComplexPathAlgorithm::complexPathsGraph(node seed, count threshold, vector<node> *adopters) {
+    const auto &g = inputGraph;
+    auto n = g.numberOfNodes();
+    Graph complex_g = Graph(n); // all nodes, no edges so far
+    vector<bool> activated(n);
+    vector<count> influence(n);
+
+    // activate threshold-1 random neighbors of seed:
+    auto needSeeds = threshold - 1;
+
+    vector<node> seeds;
+    seeds.reserve(4);
+
+    auto neighbors = Graph::NeighborRange<>(g, seed);
+
+    auto randomGen = Aux::Random::getURNG();
+    sample(neighbors.begin(), neighbors.end(), std::back_inserter(seeds), needSeeds, randomGen);
+    auto needMore = needSeeds - seeds.size();
+    if (needMore > 0) {
+        WARN("complexPathsGraph: ", seed, ": not enough direct neighbors.");
+        vector<node> moreSeeds;
+        for (auto u : neighbors) {
+            auto indirectNeighbors = Graph::NeighborRange<>(g, u);
+            for (auto in : indirectNeighbors) {
+                if (in != seed
+                    && std::find(moreSeeds.begin(), moreSeeds.end(), in) == moreSeeds.end())
+                    moreSeeds.push_back(in);
+            }
+            sample(moreSeeds.begin(), moreSeeds.end(), std::back_inserter(seeds), needMore,
+                   randomGen);
+            if (seeds.size() < needSeeds)
+                WARN("complexPathsGraph: ", seed, ": not enough indirect neighbors.");
+        }
+    }
+
+    seeds.insert(seeds.begin(), seed); // put in front, so that direct neighbors are activated first
+
+    for (auto u : seeds) {
+        activated[u] = true;
+        if (adopters)
+            adopters->push_back(u);
+        auto neighbors = Graph::NeighborRange<>(g, u);
+
+        for (auto v : neighbors) {
+            addNewEdge(complex_g, u, v);
+        }
+    }
+    // compute influences: degrees of all nodes in complex_g so far
+    auto computeInfluence = [&](node u) { influence[u] = complex_g.degree(u); };
+    complex_g.parallelForNodes(computeInfluence);
+
+    bool spread;
+    do {
+        spread = false; // at start of new iteration
+        complex_g.forNodes([&](node u) {
+            if (influence[u] >= threshold) {
+                if (!activated[u]) {
+                    activated[u] = true;
+                    if (adopters)
+                        adopters->push_back(u);
+                    spread = true;
+                }
+                auto uNeighbors = Graph::NeighborRange<>(g, u);
+                for (auto v : uNeighbors) {
+                    addNewEdge(complex_g, u, v);
+                }
+            }
```

Comment:
This results in the compiler warning: "control reaches the end of non-void function".

Reply:
Maybe in my suggestion I forgot a parenthesis. Anyway, the idea is to avoid an unnecessary indentation level. Instead of wrapping everything around `if (influence[u] >= threshold)`, you can use `if (influence[u] < threshold) return;`, the rest of the lambda remains the same. That lambda returns `void` so that compiler error is unrelated the suggestion.